// @toolpad/studio-components/dist/chunk-LHQVOIUN.js
// src/components/ErrorOverlay.tsx
import { Typography, styled, Tooltip } from "@mui/material";
import * as React from "react";
import { errorFrom } from "@toolpad/utils/errors";
import ErrorIcon from "@mui/icons-material/Error";
var OverlayRoot = styled("div")(({ theme }) => ({
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",
  padding: theme.spacing(2),
  maxWidth: "100%",
  maxHeight: "100%",
  overflow: "hidden"
}));
function ErrorContent({ sx, error }) {
  const errMessage = errorFrom(error).message;
  return /* @__PURE__ */ React.createElement(OverlayRoot, { sx }, /* @__PURE__ */ React.createElement(Typography, { sx: { display: "flex", flexDirection: "row", gap: 1 } }, /* @__PURE__ */ React.createElement(ErrorIcon, { fontSize: "small", color: "error" }), "Error"), /* @__PURE__ */ React.createElement(Tooltip, { title: errMessage }, /* @__PURE__ */ React.createElement(Typography, { variant: "body2", align: "center" }, errMessage)));
}
var ErrorOverlay = styled(ErrorContent)(({ theme }) => ({
  position: "absolute",
  inset: "0 0 0 0",
  borderWidth: 1,
  borderStyle: "solid",
  borderRadius: theme.shape.borderRadius,
  borderColor: theme.palette.divider
}));
var ErrorOverlay_default = ErrorOverlay;

export {
  ErrorContent,
  ErrorOverlay_default
};
