// @toolpad/studio-components/dist/chunk-WKPF46CV.js
import {
  createBuiltin
} from "./chunk-5BNQEU2R.js";

// src/PageRow.tsx
import * as React from "react";
import { Box } from "@mui/material";
var PageRow = React.forwardRef(function PageRow2({ layoutColumnSizes = [], gap, children, ...props }, ref) {
  const gridAutoColumns = layoutColumnSizes.reduce(
    (acc, layoutColumnSize) => `${acc}${`${acc && " "}minmax(0, ${layoutColumnSize || 1}fr)`}`,
    ""
  );
  return /* @__PURE__ */ React.createElement(
    Box,
    {
      ref,
      sx: {
        gap,
        display: "grid",
        gridAutoFlow: "column",
        gridAutoColumns
      },
      ...props
    },
    children
  );
});
var PageRow_default = createBuiltin(PageRow, {
  helperText: "A page row component.",
  argTypes: {
    gap: {
      helperText: "The gap between children.",
      type: "number",
      default: 1,
      minimum: 1
    },
    children: {
      helperText: "The content of the component.",
      type: "element",
      control: { type: "slots" }
    }
  }
});

export {
  PageRow_default
};
