// @toolpad/studio-components/dist/chunk-RBBPAYCT.js
import {
  SX_PROP_HELPER_TEXT
} from "./chunk-5TDNSO3W.js";
import {
  createBuiltin
} from "./chunk-5BNQEU2R.js";

// src/Paper.tsx
import * as React from "react";
import { Paper as MuiPaper, styled } from "@mui/material";
var PaperRoot = styled(MuiPaper)(({ theme }) => ({
  padding: theme.spacing(1),
  width: "100%"
}));
function Paper({ children, sx, ...rest }) {
  return /* @__PURE__ */ React.createElement(PaperRoot, { sx, ...rest }, children);
}
var Paper_default = createBuiltin(Paper, {
  helperText: "The Material\xA0UI [Paper](https://mui.com/material-ui/react-paper/) component.",
  layoutDirection: "vertical",
  argTypes: {
    elevation: {
      helperText: "The [elevation](https://mui.com/material-ui/react-paper/#elevation) can be used to establish a hierarchy between other content. In practical terms, the elevation controls the size of the shadow applied to the surface. In dark mode, raising the elevation also makes the surface lighter.",
      type: "number",
      minimum: 0,
      default: 1
    },
    children: {
      helperText: "The content of the component.",
      type: "element",
      control: { type: "layoutSlot" }
    },
    sx: {
      helperText: SX_PROP_HELPER_TEXT,
      type: "object"
    }
  }
});

export {
  Paper_default
};
