// @toolpad/studio-components/dist/chunk-7D3NSIRY.js
import {
  createBuiltin
} from "./chunk-5BNQEU2R.js";

// src/Tabs.tsx
import * as React from "react";
import { Tabs as MUITabs, Tab } from "@mui/material";
function Tabs({ value, onChange, tabs, defaultValue }) {
  return /* @__PURE__ */ React.createElement(
    MUITabs,
    {
      value: value || defaultValue,
      onChange: (event, newValue) => {
        onChange(newValue);
      }
    },
    tabs.map(({ title, name }) => /* @__PURE__ */ React.createElement(Tab, { label: title, value: name, key: name }))
  );
}
var Tabs_default = createBuiltin(Tabs, {
  helperText: "The Material\xA0UI [Tabs](https://mui.com/material-ui/react-tabs/) component.",
  layoutDirection: "horizontal",
  argTypes: {
    value: {
      type: "string",
      onChangeProp: "onChange",
      defaultValueProp: "defaultValue",
      helperText: "Currently active tab."
    },
    defaultValue: {
      label: "Default active tab",
      type: "string",
      default: "tab-one",
      helperText: "The tab which is active by default."
    },
    tabs: {
      type: "array",
      default: [
        {
          title: "Tab one",
          name: "tab-one"
        },
        { title: "Tab two", name: "tab-two" },
        { title: "Tab three", name: "tab-three" }
      ],
      helperText: "Tabs configuration object."
    }
  }
});

export {
  Tabs_default
};
