// @toolpad/studio/dist/exports/chunk-7HXI5QAG.mjs
// src/runtime/toolpadComponents/index.tsx
var PAGE_ROW_COMPONENT_ID = "PageRow";
var PAGE_COLUMN_COMPONENT_ID = "PageColumn";
var SPACER_COMPONENT_ID = "Spacer";
var STACK_COMPONENT_ID = "Stack";
var FORM_COMPONENT_ID = "Form";
function getElementNodeComponentId(elementNode) {
  return elementNode.attributes.component;
}
function isPageRow(elementNode) {
  return getElementNodeComponentId(elementNode) === PAGE_ROW_COMPONENT_ID;
}
function isPageColumn(elementNode) {
  return getElementNodeComponentId(elementNode) === PAGE_COLUMN_COMPONENT_ID;
}
function isPageLayoutComponent(elementNode) {
  return isPageRow(elementNode) || isPageColumn(elementNode);
}
function isFormComponent(elementNode) {
  return getElementNodeComponentId(elementNode) === FORM_COMPONENT_ID;
}

export {
  PAGE_ROW_COMPONENT_ID,
  PAGE_COLUMN_COMPONENT_ID,
  SPACER_COMPONENT_ID,
  STACK_COMPONENT_ID,
  FORM_COMPONENT_ID,
  getElementNodeComponentId,
  isPageRow,
  isPageColumn,
  isPageLayoutComponent,
  isFormComponent
};
